File: test/itemRotationSeed.mjs
const PLACEABLE_ITEM_TYPES = new Set([
  'flat-panel',
  'base',
  'counter',
  'separator',
  'sofa-set-classic',
  'sofa-single-classic',
  'sofa-double-classic',
  'coffee-table-classic',
  'table-chair-set-eames',
  'chair',
  'table-glass',
  'bar-stool',
  'mini-fridge',
  'kettle',
  'coat-rack',
  'upright',
  'profile',
  'plastic-trash-bin',
  'indoor-plant-1',
  'tv',
  'shelf',
  'led-floodlight',
  'door',
  'showcase-2',
  'showcase-3',
  'illuminated-foam',
]);

const SPECIAL = Object.freeze({
  desk_banko_100: Object.freeze({ rotationStepDeg: 45, defaultRotationDeg: 0, sideInsertRotation: 'inherit' }),
  desk_banko_150: Object.freeze({ rotationStepDeg: 45, defaultRotationDeg: 0, sideInsertRotation: 'inherit' }),
  desk_banko_200: Object.freeze({ rotationStepDeg: 45, defaultRotationDeg: 0, sideInsertRotation: 'inherit' }),
  desk_banko_100_L: Object.freeze({ rotationStepDeg: 90, defaultRotationDeg: 270, sideInsertRotation: 'inherit' }),
  desk_banko_150_L: Object.freeze({ rotationStepDeg: 90, defaultRotationDeg: 270, sideInsertRotation: 'inherit' }),
  desk_banko_200_L: Object.freeze({ rotationStepDeg: 90, defaultRotationDeg: 270, sideInsertRotation: 'inherit' }),
  furniture_sofa_single_classic: Object.freeze({
    rotationStepDeg: 45,
    defaultRotationDeg: 0,
    sideInsertRotation: 'inherit',
  }),
  furniture_bar_stool_classic: Object.freeze({
    rotationStepDeg: 45,
    defaultRotationDeg: 270,
    sideInsertRotation: 'default',
  }),
});

const DEFAULT_PLACEABLE = Object.freeze({
  rotationStepDeg: 90,
  defaultRotationDeg: 0,
  sideInsertRotation: 'inherit',
});

export function rotationFieldsForItem(itemKey, itemType) {
  if (Object.hasOwn(SPECIAL, itemKey)) return SPECIAL[itemKey];
  if (PLACEABLE_ITEM_TYPES.has(itemType)) return DEFAULT_PLACEABLE;
  return null;
}

export function applyItemRotationFields(item) {
  const fields = rotationFieldsForItem(item.itemKey, item.type);
  if (!fields) return item;
  item.rotationStepDeg = fields.rotationStepDeg;
  item.defaultRotationDeg = fields.defaultRotationDeg;
  item.sideInsertRotation = fields.sideInsertRotation;
  return item;
}

File: backend/app/modules/fair_stand/infrastructure/item_rotation_seed.py
"""Canonical Item rotation seed. Placeable modules get values; leaf rows stay null."""

from __future__ import annotations

import sqlalchemy as sa

PLACEABLE_ITEM_TYPES = frozenset(
    {
        "flat-panel",
        "base",
        "counter",
        "separator",
        "sofa-set-classic",
        "sofa-single-classic",
        "sofa-double-classic",
        "coffee-table-classic",
        "table-chair-set-eames",
        "chair",
        "table-glass",
        "bar-stool",
        "mini-fridge",
        "kettle",
        "coat-rack",
        "upright",
        "profile",
        "plastic-trash-bin",
        "indoor-plant-1",
        "tv",
        "shelf",
        "led-floodlight",
        "door",
        "showcase-2",
        "showcase-3",
        "illuminated-foam",
    }
)

_DEFAULT_PLACEABLE = (90, 0, "inherit")

_SPECIAL = {
    "desk_banko_100": (45, 0, "inherit"),
    "desk_banko_150": (45, 0, "inherit"),
    "desk_banko_200": (45, 0, "inherit"),
    "desk_banko_100_l": (90, 270, "inherit"),
    "desk_banko_150_l": (90, 270, "inherit"),
    "desk_banko_200_l": (90, 270, "inherit"),
    "furniture_sofa_single_classic": (45, 0, "inherit"),
    "furniture_bar_stool_classic": (45, 270, "default"),
}


def rotation_fields_for_item(item_key: str, item_type: str) -> tuple[int | None, int | None, str | None]:
    if item_key in _SPECIAL:
        return _SPECIAL[item_key]
    if item_type in PLACEABLE_ITEM_TYPES:
        return _DEFAULT_PLACEABLE
    return (None, None, None)


def apply_item_rotation_fields(row: dict) -> dict:
    step, default, side = rotation_fields_for_item(row["item_key"], row["item_type"])
    row["rotation_step_deg"] = step
    row["default_rotation_deg"] = default
    row["side_insert_rotation"] = side
    return row


def fill_item_rotation_columns(bind) -> None:
    existing = bind.execute(sa.text("SELECT item_key, item_type FROM fair_stand_items")).mappings()
    for row in existing:
        step, default, side = rotation_fields_for_item(row["item_key"], row["item_type"])
        bind.execute(
            sa.text(
                "UPDATE fair_stand_items SET rotation_step_deg = :step, "
                "default_rotation_deg = :default_deg, side_insert_rotation = :side "
                "WHERE item_key = :item_key"
            ),
            {
                "step": step,
                "default_deg": default,
                "side": side,
                "item_key": row["item_key"],
            },
        )
